File: src/main/java/io/github/tanguygab/playerlistexpansion/filters/Filter.java
package io.github.tanguygab.playerlistexpansion.filters;

import io.github.tanguygab.playerlistexpansion.PlayerListExpansion;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public abstract class Filter {

    @SuppressWarnings("all")
    private static final Map<String,Function<String,Filter>> filters = new HashMap<String,Function<String,Filter>>() {{
        put("BANNED",arg->new Banned());
        put("CANSEE",arg->new CanSee());
        put("GAMEMODE",GameMode::new);
        put("NEARBY",Nearby::new);
        put("PERMISSION",Permission::new);
        put("PLACEHOLDER",Placeholder::new);
        if (Bukkit.getServer().getPluginManager().isPluginEnabled("ViaVersion"))
            put("VERSION",Version::new);
        put("WHITELISTED",arg->new Whitelisted());
        put("WORLD",World::new);
    }};
    public static Filter find(String string) {
        int index = string.indexOf(":");
        String filter = index == -1 ? string : string.substring(0,index);
        filter = filter.toUpperCase();
        String arg = index == -1 ? null : string.substring(index+1);
        return filters.containsKey(filter) ? filters.get(filter).apply(arg) : null;
    }

    protected String[] split(String arg) {
        return arg.split(PlayerListExpansion.get().argumentSeparator);
    }

    protected OfflinePlayer getOffline(String name) {
        for (OfflinePlayer player : Bukkit.getServer().getOfflinePlayers())
            if (name.equals(player.getName()))
                return player;
        return null;
    }
    protected Player getOnline(String name) {
        return Bukkit.getServer().getPlayerExact(name);
    }

    public abstract boolean filter(String name, OfflinePlayer viewer);

}

File: src/main/java/io/github/tanguygab/playerlistexpansion/filters/Filter.kt
package io.github.tanguygab.playerlistexpansion.filters

import io.github.tanguygab.playerlistexpansion.PlayerListExpansion
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import java.util.*
import java.util.function.Function

abstract class Filter {
    var isInverted: Boolean = false
        private set

    protected fun split(arg: String): List<String> {
        return arg.split(PlayerListExpansion.get().argumentSeparator)
    }

    protected fun getOffline(name: String?): OfflinePlayer? {
        return if (name === null) null else Bukkit.getServer().getOfflinePlayerIfCached(name)
    }

    protected fun getOnline(name: String?): Player? {
        return if (name === null) null else Bukkit.getServer().getPlayerExact(name)
    }

    abstract fun filter(name: String?, viewer: OfflinePlayer?): Boolean

    companion object {
        private val filters = HashMap<String, Function<String?, Filter>>().apply {
            put("BANNED") { Banned() }
            put("CANSEE") { CanSee() }
            put("GAMEMODE") { GameMode(it) }
            put("NEARBY") { Nearby(it!!) }
            put("PERMISSION") { Permission(it!!) }
            put("PLACEHOLDER") { Placeholder(it!!) }
            if (Bukkit.getServer().pluginManager.isPluginEnabled("ViaVersion"))
                put("VERSION") { Version(it!!) }
            put("WHITELISTED") { Whitelisted() }
            put("WORLD") { World(it!!) }
        }

        fun find(string: String): Filter? {
            val args = string.split(":", limit = 2)

            val inverted = args[0][0] == '!'
            val filter = args[0].uppercase().substring(if (inverted) 1 else 0)
            val arg = if (args.size > 1) args[1] else null

            if (filters.containsKey(filter)) {
                val f = filters[filter]!!.apply(arg)
                f.isInverted = inverted
                return f
            }
            return null
        }
    }
}
